### ISSUE TITLE
Mutation result never gets past isLoading

### ISSUE THREAD

**Issue by derochedesign** [reduxjs/redux-toolkit#2910]
I've been working at this one for a while and haven't got anywhere productive. I have a mutation to create a post, which requires a large media attached to it. When the user clicks "post", the media is uploaded via axios (to provide progress). Once that upload is finished, the `createPost` mutation is called. 

Now the post is always successfully created. But in that instance of the browser, the `postResult` useEffect never sees an update for `isSuccess` and is stuck in `isLoading`. Once I refresh the browser, the new post is there properly.

I read through this post: https://github.com/reduxjs/redux-toolkit/issues/2167 , but I don't see much connection.

axios `upload` function (for media):
```js
export async function upload(
  data: any,
  id: string,
  auth: any,
  api: string,
  controller: any,
  progressCallback: any,
  abortCallback: any,
  completedCallback: any,
) {
  const url = "url here";

  const config = {
    headers: {
      Authorization: `Bearer ${auth.access_token}`,
      'Content-Type': 'multipart/form-data',
    },
    //TODO: debounce (500 or 1000)
    onUploadProgress: (progressEvent: any) =>
      progressCallback(progressEvent, id),
    signal: controller.signal,
  };

  await axios
    .post(url, data, config)
    .then((res: any) => {
      completedCallback(res);
    })
    .catch((e: any) => {
      console.error('User aborted upload:', e);
      abortCallback(id);
    });
}
```

`completedCallback(res)` then calls `createUserPost`:

```js
const createUserPost = (res: any) => {
    const postObj = {
      content: (uploadMetaDataRef.current as any)?.generalProjectTitle,
      media: [
        (res as any).data._id,
        uploadImageResultRef.current.data._id,
      ],
    };
    //postObj is properly populated/configured at this point
    createPost(postObj);
 }
```

And finally the `useEffect`:

```js
useEffect(() => {
  if (postResult.isError || postResult.isSuccess) {
    //never gets here
    setShowUploadInfo(true);
    setIsUploading(false);
  }
}, [postResult]);
```

The final postResult I get is:
```json
{
  "requestId": "a1E9XToUbnrdA3QXH1Eoa",
  "status": "pending",
  "endpointName": "createPost",
  "startedTimeStamp": 1668499076779,
  "isUninitialized": false,
  "isLoading": true,
  "isSuccess": false,
  "isError": false,
  "originalArgs": {
    "content": "song96",
    "media": [
      "63734684c8183c9f326c0e7a",
      "63734684c8183c9f326c0e72"
    ],
  }
}
```

This is all in a react context if that means anything.

**Comment by phryneas** [comment]
You pretty much emitted all code that would show how you are using RTK Query. Could you please add the `createApi` call with the relevant endpoint as well as show how you call the mutation hook in your component?

**Comment by phryneas** [comment]
I still fail to see how any of that mutation is ever calling your `upload` function in any way.  
How are those related?  
Why are you talking about a mutation result that is stuck in `isLoading` and share completely unrelated code?

I'm sorry, but I'm seriously confused.

**Comment by derochedesign** [comment]
The chain of events are: user clicks post -> axios media upload starts (`upload` function) -> axios media upload finishes and calls `createUserPost` with relevant response -> `createUserPost` calls the `createPost` mutation -> postResult _should_ be successful. 

AKA, the `upload` function calls the mutation once complete, not the other way around.

When the user clicks "post" it calls
```js
const StartPostMediaUpload = async (data: any, api: string) => {
    if (isUploading) return;
    setIsUploading(true);
    setUploadMetadata(data.metadata);

    await uploadImage(data.image);
    await upload(
      data.audio,
      data.audio.title,
      auth,
      api,
      controller.current,
      updateProgress,
      abortUpload,
      createUserPost,
      false
    );
  };
```
The `await upload()` here is where the connection between `upload` and `createUserPost` is. I give `upload` both the media to upload and the `completedCallback` (which is `createUserPost`). When the given media is uploaded (via axios), it should give the response of that to `completedCallback`/`createUserPost` where it should then create the proper `postObj` and post it via the RTK mutation.

**Comment by phryneas** [comment]
So the only relevant code for this is essentially how `createPost` is defined, and how you call `createPost`.

Can you **please** share the component where all of this happens? How you call `createPost`? How you call the `useEffect`? How you call the mutation hook?

And I mean the full working component.  
It's fine if you remove stuff, but **remove that in your code** and see if it still does exactly the same thing. Don't omit anything beyond that.  
I cannot help you at all if I get bits and pieces without any missing glue. I need that code that runs on your machine and triggers the error.

Honestly, me asking you for this is already a very big compromise.  
In that other issue I linked I am asking for a full reproduction again and again - and so far people just complain and nobody has ever followed up with something I can run on my own machine and actually run.  
I don't really know how to help people with this.  
It's like the 20th person in a row is coming to the car repair shop describing that it is not driving, but refusing to bring in the car for maintenance.

**Comment by derochedesign** [comment]
If how `createPost` is defined and called is the only relevant code then it would be working, because all other instances of this setup work for me without issue. I figure it must be to do with the lead up to the call/how the entire chain of events happens so that's why I included it all.

I figured that I was just missing something obvious that someone would catch easily. I will get a reproducable example for you later in the day.

Either way, thanks for putting up with all of it, your help doesn't go unnoticed.
